server/lomas_server/dp_queries/dp_libraries/opendp.py

```diff
@@ -152,20 +398,64 @@ def has_dataset_input_metric(pipeline: dp.Measurement) -> None:
         raise InvalidQueryException(e)
 
 
-def reconstruct_measurement_pipeline(pipeline: str) -> dp.Measurement:
+def extract_group_by_columns(plan: str) -> list | None:
+    """
+    Extract column names used in the BY operation from the plan string.
+    Parameters:
+    plan (str): The polars query plan as a string.
+    Returns:
+    list: A list of column names used in the BY operation.
+    """
+    # Regular expression to capture the content inside BY []
+    aggregate_by_pattern = r"AGGREGATE(?:.|\n)+?BY \[(.*?)\]"
+
+    # Find the part of the plan related to the GROUP BY clause
+    match = re.search(aggregate_by_pattern, plan)
+
+    if match:
+        # Extract the columns part
+        columns_part = match.group(1)
+        # Find all column names inside col("...")
+        column_names = re.findall(r'col\("([^"]+)"\)', columns_part)
+        return column_names
+    return None
+
+
+def reconstruct_measurement_pipeline(query_json: OpenDPQueryModel, metadata: dict) -> dp.Measurement:
     """Reconstruct OpenDP pipeline from json representation.
 
     Args:
-        pipeline (str): The JSON string encoding of the pipeline.
+        query_json (BaseModel): The JSON request object for the query.
+        metadata (dict): The dataset metadata dictionary.\
+            Only used for polars pipelines.
 
     Raises:
-        InvalidQueryException: If the pipeline is not a measurement.
+        InvalidQueryException: If the pipeline is not a measurement or\
+            the pipeline type is not supported.
 
     Returns:
         dp.Measurement: The reconstructed pipeline.
     """
     # Reconstruct pipeline
-    opendp_pipe = make_load_json(pipeline)
+    if query_json.pipeline_type == "legacy":
+        opendp_pipe = make_load_json(query_json.opendp_json)
+    elif query_json.pipeline_type == "polars":
+        # TODO Might pickle, huge security implications!!
+        plan = pl.LazyFrame.deserialize(io.StringIO(query_json.opendp_json), format="json")
+
+        groups = extract_group_by_columns(plan.explain())
+        output_measure = {
+            "laplace": dp.measures.max_divergence(),
+            "gaussian": dp.measures.zero_concentrated_divergence(),
+        }[query_json.mechanism]
```

Comment:
I think here you get less mypy-trouble if mechanism is a complete type and you can match here otherwise ...[tpe] where tpe can be None is always an issue